### DISCUSSION TITLE
blackbear submodule not linking against conda std library

### DISCUSSION THREAD

**Discussion by lynnmunday** [idaholab/moose#16867] [Q&A Getting Started] [answered]
I'm getting a build error on my Mac laptop when I try to build blackbear with the neml submodule.  I can build blackbear without neml and I can build neml by itself but I keep getting the below error when I try to build blackbear with the neml submodule.  I can build them together on sawtooth so it must be something in my environment but I'm not sure what.  I've deleted my moose repo, blackbear repo and re-cloned them.  I'm using the standard moose conda environment using the instructions from the moose conda install help page. I brew uninstalled everything to make sure that wasn't causing the conflict and it wasn't because I still get the neml build error.  I think there is something in the blackbear/contrib/neml.mk that is causing this but I can't figure out what.  Has anyone seen this error and know how to fix it?

```
In file included from /Users/mundlb/projects/blackbear/contrib/neml/src/Block.h:3:
In file included from /Users/mundlb/projects/blackbear/contrib/neml/src/models.h:4:
In file included from /Users/mundlb/projects/blackbear/contrib/neml/src/objects.h:6:
In file included from /Users/mundlb/miniconda3/envs/moose-noLibmesh/bin/../include/c++/v1/vector:276:
/Users/mundlb/miniconda3/envs/moose-noLibmesh/bin/../include/c++/v1/__bit_reference:172:38: error: no member named 'min' in namespace 'std::__1'
        __storage_type __dn = _VSTD::min(__clz_f, __n);
```

**Comment by lindsayad** [comment]
Best match I've found so far is: https://trac.macports.org/ticket/60962

**Reply by lindsayad** [reply]
It would be good to check the include lines to make sure at least the C++ includes are not different for your different builds (e.g. neml alone, blackbear alone, blackbear + neml)

**Reply by lynnmunday** [reply]
I don't see any differences with the includes other than the extra neml dependencies. I followed the instructions at the link you posted but that doesn't solve the problem either.  I'm still picking up the wrong std libraries.  The comments in the answer to this on stackoverflow https://stackoverflow.com/questions/29293394/where-does-the-1-symbol-come-from-when-using-llvms-libc#:~:text=The%20LLVM%20C%2B%2B%20runtime%20(%20libc,(like%20std%3A%3Astring%20).   explain the problem but I don't see what the solution to it is.  It looks like the neml header file objects.h is #including \<algorithm\> and my interpretation of the error I posted, it is trying to link to the libc++ version that has the __1 decoration, std::__1::min, but it cannot find this function in algorithms version that I have in my conda environment which is the libstdc++ version that is std::min function.

**Reply by lindsayad** [reply]
can you paste your include directories?

**Reply by lynnmunday** [reply]
This is my make -n output for one neml source file:

echo "Compiling C++ (in "opt" mode) "/Users/mundlb/projects/blackbear/contrib/neml/src/nemlerror.cxx"..."
/Users/mundlb/miniconda3/envs/moose/libmesh/contrib/bin/libtool  --tag=CXX  --mode=compile --quiet \
	  mpicxx -DNDEBUG -DADFPARSER_INCLUDES="\"-I/Users/mundlb/projects/moose/framework/include/utils -I/Users/mundlb/projects/moose/framework/include/base -I/Users/mundlb/miniconda3/envs/moose/libmesh/include\"" -DHEAT_CONDUCTION_ENABLED -DMISC_ENABLED -DTENSOR_MECHANICS_ENABLED -DSTOCHASTIC_TOOLS_ENABLED -DXFEM_ENABLED -DCONTACT_ENABLED -DMODULE_LOADER_ENABLED -DNEML_ENABLED -DNEML_STRAIN_RATE_LIMIT=1e10 -DBLACKBEAR_ENABLED -O2 -felide-constructors -Qunused-arguments -Wunused-parameter -Wunused -fopenmp -DHAVE_GPERFTOOLS -I/Users/mundlb/code/gperftools/installed/include  -DMETHOD=opt -Werror=return-type -Werror=reorder -Woverlength-strings -I/Users/mundlb/projects/blackbear/contrib/neml/rapidxml -I/Users/mundlb/projects/blackbear/contrib/neml_extra_include  -I/Users/mundlb/projects/moose/framework/build/header_symlinks  -I/Users/mundlb/projects/moose/framework/contrib/boost/include  -I/Users/mundlb/projects/moose/framework/contrib/json/include  -I/Users/mundlb/projects/moose/framework/contrib/json/include/nlohmann  -I/Users/mundlb/projects/moose/framework/contrib/mtwist/include  -I/Users/mundlb/projects/moose/framework/contrib/pcre/include  -I/Users/mundlb/projects/moose/framework/contrib/pugixml/include  -I/Users/mundlb/projects/moose/framework/contrib/tinydir/include  -I"/Users/mundlb/projects/moose/framework/contrib/gtest"  -I"/Users/mundlb/projects/moose/framework/contrib/hit" -I/Users/mundlb/miniconda3/envs/moose/libmesh/include -I/Users/mundlb/miniconda3/envs/moose/libmesh-vtk/include/vtk-6.3 -D_THREAD_SAFE -pthread -I/Users/mundlb/miniconda3/envs/moose/include -I/Users/mundlb/miniconda3/envs/moose//include -I/Users/mundlb/projects/moose/modules/heat_conduction/build/header_symlinks  -I/Users/mundlb/projects/moose/modules/misc/build/header_symlinks  -I/Users/mundlb/projects/moose/modules/tensor_mechanics/build/header_symlinks  -I/Users/mundlb/projects/moose/modules/stochastic_tools/build/header_symlinks  -I/Users/mundlb/projects/moose/modules/xfem/build/header_symlinks  -I/Users/mundlb/projects/moose/modules/contact/build/header_symlinks  -I/Users/mundlb/projects/moose/modules/module_loader/build/header_symlinks  -I/Users/mundlb/projects/blackbear/build/header_symlinks -I/Users/mundlb/projects/blackbear/contrib/neml/src -I/Users/mundlb/projects/blackbear/contrib/neml/rapidxml -I/Users/mundlb/projects/blackbear/contrib/neml_extra_include -I/Users/mundlb/miniconda3/envs/moose/libmesh/include -I/Users/mundlb/miniconda3/envs/moose/libmesh-vtk/include/vtk-6.3 -D_THREAD_SAFE -pthread -I/Users/mundlb/miniconda3/envs/moose/include -I/Users/mundlb/miniconda3/envs/moose//include -w -DHAVE_CONFIG_H -MMD -MP -MF /Users/mundlb/projects/blackbear/contrib/neml/src/nemlerror.x86_64-apple-darwin13.4.0.opt.lo.d -MT /Users/mundlb/projects/blackbear/contrib/neml/src/nemlerror.x86_64-apple-darwin13.4.0.opt.lo -c /Users/mundlb/projects/blackbear/contrib/neml/src/nemlerror.cxx -o /Users/mundlb/projects/blackbear/contrib/neml/src/nemlerror.x86_64-apple-darwin13.4.0.opt.lo
echo "Compiling C++ (in "opt" mode) "/Users/mundlb/projects/blackbear/contrib/neml/src/surfaces.cxx"..."

**Reply by lynnmunday** [reply]
If I paste in that command and add -H after mpicxx I get the following:  (this is only a little bit of it) It is weird because it mostly includes the std library from miniconda but then also includes the std lib from CommandLineTools for instance in these two for string.h
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/string.h
......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/string.h


. /Users/mundlb/projects/blackbear/contrib/neml/src/visco_flow.h
.. /Users/mundlb/projects/blackbear/contrib/neml/src/objects.h
... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/string
.... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
.... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/string_view
..... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
..... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__string
...... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
...... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/algorithm
....... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
....... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/initializer_list
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/cstddef
......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/version
.......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
......... /Users/mundlb/miniconda3/envs/moose/lib/clang/11.0.0/include/stddef.h
.......... /Users/mundlb/miniconda3/envs/moose/lib/clang/11.0.0/include/__stddef_max_align_t.h
......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__nullptr
.......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
....... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/type_traits
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
....... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/cstring
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/string.h
......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/string.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_types.h
........... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types.h
............ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_symbol_aliasing.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_posix_availability.h
............ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/_types.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/i386/_types.h
............ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_pthread/_pthread_types.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/Availability.h
........... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/AvailabilityInternal.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_size_t.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_null.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_rsize_t.h
........... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/types.h
............ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/i386/types.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int8_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int16_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int32_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_int64_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int8_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int16_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int32_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int64_t.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_intptr_t.h
.............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/types.h
............. /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_uintptr_t.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_errno_t.h
.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_ssize_t.h
**.......... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/strings.h**
........... /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_size_t.h
**........... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/string.h**
....... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/utility
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__tuple
......... /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/__config
........ /Users/mundlb/miniconda3/envs/moose/bin/../include/c++/v1/cstdint

**Reply by lindsayad** [reply]
I assume that you are building within your conda environment. Are you using your own built libmesh or libmesh from conda? I am curious what happens if you try to `make` from blackbear after you apply this patch to `blackbear`:
```
diff --git a/contrib/neml.mk b/contrib/neml.mk
index 10facd5..8b134ae 100644
--- a/contrib/neml.mk
+++ b/contrib/neml.mk
@@ -21,7 +21,7 @@ $(neml_LIB): $(neml_objects)
 $(NEML_DIR)/src/%.$(obj-suffix) : $(NEML_DIR)/src/%.cxx
 	@echo "Compiling C++ (in "$(METHOD)" mode) "$<"..."
 	@$(libmesh_LIBTOOL) --tag=CXX $(LIBTOOLFLAGS) --mode=compile --quiet \
-	  $(libmesh_CXX) $(libmesh_CPPFLAGS) $(ADDITIONAL_CPPFLAGS) $(libmesh_CXXFLAGS) $(neml_dep_includes) $(app_INCLUDES) $(libmesh_INCLUDE) -w -DHAVE_CONFIG_H -MMD -MP -MF $@.d -MT $@ -c $< -o $@
+	  $(libmesh_CXX) $(libmesh_CPPFLAGS) $(ADDITIONAL_CPPFLAGS) $(CXXFLAGS) $(libmesh_CXXFLAGS) $(neml_dep_includes) $(app_INCLUDES) $(libmesh_INCLUDE) -w -DHAVE_CONFIG_H -MMD -MP -MF $@.d -MT $@ -c $< -o $@
 
 ADDITIONAL_INCLUDES  += -I$(neml_includes) $(neml_dep_includes)
 ADDITIONAL_LIBS      += -L$(NEML_DIR) -lneml-$(METHOD)
```

**Reply by lindsayad** [reply]
This is just adding `$(CXXFLAGS)` into your compile line

**Reply by lindsayad** [reply]
So I just activated a conda `moose-libmesh` environment and started compiling and got this error:
```
In file included from /Users/lindad/projects/moose/framework/contrib/pcre/src/pcre_stringpiece.cc:37:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/iostream:37:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/ios:215:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/__locale:14:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/string:506:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/string_view:175:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/__string:57:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/algorithm:644:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/functional:512:
In file included from /Users/lindad/projects/blackbear/contrib/neml/src/Block.h:3:
In file included from /Users/lindad/projects/blackbear/contrib/neml/src/models.h:4:
In file included from /Users/lindad/projects/blackbear/contrib/neml/src/objects.h:5:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/sstream:173:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/ostream:139:
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/streambuf:141:5: error: unknown type name 'locale'
    locale pubimbue(const locale& __loc) {
    ^
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/streambuf:141:27: error: unknown type name 'locale'
    locale pubimbue(const locale& __loc) {
                          ^
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/streambuf:149:5: error: unknown type name 'locale'
    locale getloc() const { return __loc_; }
    ^
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/streambuf:153:48: error: unknown type name 'streamsize'
    basic_streambuf* pubsetbuf(char_type* __s, streamsize __n)
                                               ^
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/streambuf:157:41: error: incomplete type 'std::__1::ios_base' named in nested name specifier
    pos_type pubseekoff(off_type __off, ios_base::seekdir __way,
                                        ^~~~~~~~~~
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/iosfwd:103:24: note: forward declaration of 'std::__1::ios_base'
class _LIBCPP_TYPE_VIS ios_base;
```
and later on, one that looks a lot like yours!
```
In file included from /Users/lindad/projects/moose/framework/contrib/pcre/src/pcrecpp.cc:43:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/string:506:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/string_view:175:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/__string:57:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/algorithm:644:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/functional:512:
In file included from /Users/lindad/projects/blackbear/contrib/neml/src/Block.h:3:
In file included from /Users/lindad/projects/blackbear/contrib/neml/src/models.h:4:
In file included from /Users/lindad/projects/blackbear/contrib/neml/src/objects.h:5:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/sstream:173:
In file included from /Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/ostream:139:
/Users/lindad/miniconda3/envs/moose-libmesh/bin/../include/c++/v1/streambuf:412:45: error: no member named 'min' in namespace 'std::__1'
            const streamsize __len = _VSTD::min(static_cast<streamsize>(INT_MAX),
```

**Reply by lindsayad** [reply]
The stack trace is saying it all: line 512 of the c++ header `functional` is including `Block.h` which just so happens to be a neml header. Obviously the STL headers did not mean to include a header from NEML. This is why projects like libmesh make their includes like so: `#include "libmesh/foo.h"` to avoid any naming clashes. So I would tell the maintainers of NEML to rename their header or change how NEML headers are meant to be included, e.g. `#include "neml/Block.h"`. (Note that MOOSE is guilty of not guarding against name clashes as well)

**Reply by lindsayad** [reply]
The thing I have to figure out is why NEML doesn't itself have issues when it's built standalone. So the fault may very well not be with NEML but instead with the flags we're using when compiling blackbear with NEML present.

**Comment by lindsayad** [comment, marked as answer]
Ok I think I've figured this out. Here is a test program:
```c++
#include <iostream>
#include <some_crazy_name_bar>
using namespace std;

int main()
{
  bar();
  cout << "Hello World!\nI'm a C++ program\n";
}
```
and here is the directory containing it:
```
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ ls
a.out               hello.cpp           iostream            some_crazy_name_bar
```
`iostream` is some dummy file that that does not contain a definition of `std::cout`. `some_crazy_name_bar` is a file that defines the method `void bar() {}`. Here are some different compilation attempts:
```
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ clang++ -I. hello.cpp 
hello.cpp:4:17: warning: using directive refers to implicitly-defined namespace 'std'
using namespace std;
                ^
hello.cpp:8:2: error: use of undeclared identifier 'cout'
        cout << "Hello World!\nI'm a C++ program\n";
        ^
1 warning and 1 error generated.
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ clang++ hello.cpp
hello.cpp:4:10: error: 'some_crazy_name_bar' file not found with <angled> include; use "quotes" instead
#include <some_crazy_name_bar>
         ^~~~~~~~~~~~~~~~~~~~~
         "some_crazy_name_bar"
1 error generated.
```
And then following clang's suggestion to use quotes for including `some_crazy_name_bar`, the program compiles fine:
```
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ clang++ hello.cpp
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ ./a.out
Hello World!
I'm a C++ program
```
Another compilation method that works fine:
```
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ clang++ -iquote. hello.cpp
(moose-libmesh) lindad@localhost:~/programming/cpp/test-includes(master)$ ./a.out
Hello World!
I'm a C++ program
```
So the key takeaways:

1. `-I<dir>` adds `<dir>` to the search path for both quoted and angle-bracket included files. *This* is what destroys blackbears compilation of NEML because it does: `-I${BLACKBEAR_DIR}/contrib/neml/src` and consequently all of the headers in there become candidates for both quoted and angle-bracket includes. *This* is why when compiling blackbear we pick up the `block.h` header as a system header. When NEML is building itself it *does not* `-I${BLACKBEAR_DIR}/contrib/neml/src` and so it does fine
2. `-iquote<dir>` adds `<dir>` only to the search path for quoted includes. *This* is what we are going to want to do for sustainable builds in MOOSE and Blacbear. I actually have to check and see whether this is an option for both gcc and clang. All the research I've done here for this discussion has been with clang.

I am honestly kind of shocked we haven't run into an issue like this before

**Reply by lynnmunday** [reply]
THANKS @lindsayad   This should be on stack exchange.  No one had an explanation like this.  
I've replaced ```-I``` with ```-iquote``` in blackbear/contrib/neml.mk and it is building fine.
Is this the fix you would recommend?  I can push it to a pr and see if it passes all of the tests, those should be testing with both clang and gcc.  Or should I change all of the headers in NEML to something like "src/block.h" ?

**Reply by lindsayad** [reply]
You're welcome! I realized that my "Another compilation method that works fine" didn't actually show the `-iquote` compilation example. I've updated the answer to have it.

**Reply by lynnmunday** [reply]
I'm going to make an issue for this in blackbear and update the neml.mk file with the ```-I``` replaced with ```-iquote``` .  I tried this on sawtooth which loads the gcc compilers and it still builds.

**Reply by lindsayad** [reply]
That's great to hear. No I think the `-I` -> `-iquote` fix is sufficient. `src/foo` looks kind of ugly and the `src/` prefix is also pretty generic so you could still have name clashes. And yea I checked that gcc accepts `-iquote` as well, so I definitely think this is a great fix and something we will want to do more across MOOSE builds

**Comment by dschwen** [comment]
D'Uh, Lynn, you're encountering this because you're using a case
insensitive filesystem. My fs is case sensitive (as I'm used to that from
Linux), so I was unable to reproduce this issue.

**Reply by lindsayad** [reply]
I know case insensitive... smh. But I was also thinking about that and was wondering why there was no warning about non-portable paths since the system header is `Block.h` but the NEML file is `block.h`. I checked and there were no warnings in my blackbear compilation before the hundreds of errors. I do get a warning in my toy example if I change `iostream` to `Iostream`
